### DISCUSSION TITLE
Get only preds for one class in RTMPose (same config used as rtm ap10k model)

### DISCUSSION THREAD

**Discussion by NicolasPetermann134** [open-mmlab/mmpose#2444] [Q&A] [answered]
Hi there, 

I trained a model with 3 Classes it has similair format as ap10k. Therefore I used the config file from ap10k and adjusted only batch size and name of dataset, paths etc. :

_base_ = ['/home/nico/PycharmProjects/mmpose/configs/_base_/default_runtime.py']
import torch

#load custom_keys
custom_keys = torch.load('/home/nico/PycharmProjects/mmpose/privat_configs/custom_keys.pth')
# runtime
max_epochs = 420
stage2_num_epochs = 30
base_lr = 4e-3

train_cfg = dict(max_epochs=max_epochs, val_interval=10)
randomness = dict(seed=21)

# optimizer
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=dict(type='AdamW', lr=base_lr, weight_decay=0.05),
    paramwise_cfg=dict(
        #custom_keys,
        norm_decay_mult=0, bias_decay_mult=0, bypass_duplicate=True))

# learning rate
param_scheduler = [
    dict(
        type='LinearLR',
        start_factor=1.0e-5,
        by_epoch=False,
        begin=0,
        end=1000),
    dict(
        # use cosine lr from 150 to 300 epoch
        type='CosineAnnealingLR',
        eta_min=base_lr * 0.05,
        begin=max_epochs // 2,
        end=max_epochs,
        T_max=max_epochs // 2,
        by_epoch=True,
        convert_to_iter_based=True),
]

# automatically scaling LR based on the actual training batch size
auto_scale_lr = dict(base_batch_size=64)

# codec settings
codec = dict(
    type='SimCCLabel',
    input_size=(256, 256),
    sigma=(5.66, 5.66),
    simcc_split_ratio=2.0,
    normalize=False,
    use_dark=False)
#resume = True
#load_from = '/home/nico/PycharmProjects/mmpose/work_dirs/rtmpose-m_simcc-aic-coco_pt-aic-coco_420e-256x192-63eb25f7_20230126.pth'
#load_from = "/home/nico/PycharmProjects/mmpose/tools/work_dirs/rtmpose-m_8xb64-210e-DG_dataset_01-256x256/epoch_210.pth"
# model settings
model = dict(
    type='TopdownPoseEstimator',
    data_preprocessor=dict(
        type='PoseDataPreprocessor',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        bgr_to_rgb=True),
    backbone=dict(
        _scope_='mmdet',
        type='CSPNeXt',
        arch='P5',
        expand_ratio=0.5,
        deepen_factor=0.67,
        widen_factor=0.75,
        out_indices=(4, ),
        channel_attention=True,
        norm_cfg=dict(type='SyncBN'),
        act_cfg=dict(type='SiLU'),
        init_cfg=dict(
            type='Pretrained',
            prefix='backbone.',
            checkpoint='https://download.openmmlab.com/mmpose/v1/projects/'
            'rtmpose/cspnext-m_udp-aic-coco_210e-256x192-f2f7d6f6_20230130.pth'  # noqa
        )),
    head=dict(
        type='RTMCCHead',
        in_channels=768,
        out_channels=17,
        input_size=codec['input_size'],
        in_featuremap_size=(8, 8),
        simcc_split_ratio=codec['simcc_split_ratio'],
        final_layer_kernel_size=7,
        gau_cfg=dict(
            hidden_dims=256,
            s=128,
            expansion_factor=2,
            dropout_rate=0.,
            drop_path=0.,
            act_fn='SiLU',
            use_rel_bias=False,
            pos_enc=False),
        loss=dict(
            type='KLDiscretLoss',
            use_target_weight=True,
            beta=10.,
            label_softmax=True),
        decoder=codec),
    test_cfg=dict(flip_test=True, ))

# base dataset settings
dataset_type = 'AP10KDataset'
data_mode = 'topdown'
data_root = '/home/nico/PycharmProjects/opencv/yolov7_multiclass/DG_dataset_01_/'

backend_args = dict(backend='local')
# backend_args = dict(
#     backend='petrel',
#     path_mapping=dict({
#         f'{data_root}': 's3://openmmlab/datasets/pose/ap10k/',
#         f'{data_root}': 's3://openmmlab/datasets/pose/ap10k/'
#     }))

# pipelines
train_pipeline = [
    dict(type='LoadImage', backend_args=backend_args),
    dict(type='GetBBoxCenterScale'),
    dict(type='RandomFlip', direction='horizontal'),
    dict(type='RandomHalfBody'),
    dict(
        type='RandomBBoxTransform', scale_factor=[0.6, 1.4], rotate_factor=80),
    dict(type='TopdownAffine', input_size=codec['input_size']),
    dict(type='mmdet.YOLOXHSVRandomAug'),
    dict(
        type='Albumentation',
        transforms=[
            dict(type='Blur', p=0.1),
            dict(type='MedianBlur', p=0.1),
            dict(
                type='CoarseDropout',
                max_holes=1,
                max_height=0.4,
                max_width=0.4,
                min_holes=1,
                min_height=0.2,
                min_width=0.2,
                p=1.0),
        ]),
    dict(type='GenerateTarget', encoder=codec),
    dict(type='PackPoseInputs')
]
val_pipeline = [
    dict(type='LoadImage', backend_args=backend_args),
    dict(type='GetBBoxCenterScale'),
    dict(type='TopdownAffine', input_size=codec['input_size']),
    dict(type='PackPoseInputs')
]

train_pipeline_stage2 = [
    dict(type='LoadImage', backend_args=backend_args),
    dict(type='GetBBoxCenterScale'),
    dict(type='RandomFlip', direction='horizontal'),
    dict(type='RandomHalfBody'),
    dict(
        type='RandomBBoxTransform',
        shift_factor=0.,
        scale_factor=[0.75, 1.25],
        rotate_factor=60),
    dict(type='TopdownAffine', input_size=codec['input_size']),
    dict(type='mmdet.YOLOXHSVRandomAug'),
    dict(
        type='Albumentation',
        transforms=[
            dict(type='Blur', p=0.1),
            dict(type='MedianBlur', p=0.1),
            dict(
                type='CoarseDropout',
                max_holes=1,
                max_height=0.4,
                max_width=0.4,
                min_holes=1,
                min_height=0.2,
                min_width=0.2,
                p=0.5),
        ]),
    dict(type='GenerateTarget', encoder=codec),
    dict(type='PackPoseInputs')
]

# data loaders
train_dataloader = dict(
    batch_size=64,
    num_workers=10,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_mode=data_mode,
        ann_file='annotations/annotations_coco_train.json',
        data_prefix=dict(img='data/'),
        pipeline=train_pipeline,
    ))
val_dataloader = dict(
    batch_size=16,
    num_workers=10,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False, round_up=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_mode=data_mode,
        ann_file='annotations/annotations_coco_valid.json',
        data_prefix=dict(img='data/'),
        test_mode=True,
        pipeline=val_pipeline,
    ))
test_dataloader = dict(
    batch_size=16,
    num_workers=10,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False, round_up=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_mode=data_mode,
        ann_file='annotations/annotations_coco_valid.json',
        data_prefix=dict(img='data/'),
        test_mode=True,
        pipeline=val_pipeline,
    ))

# hooks
default_hooks = dict(
    checkpoint=dict(save_best='coco/AP', rule='greater', max_keep_ckpts=1))

custom_hooks = [
    dict(
        type='EMAHook',
        ema_type='ExpMomentumEMA',
        momentum=0.0002,
        update_buffers=True,
        priority=49),
    dict(
        type='mmdet.PipelineSwitchHook',
        switch_epoch=max_epochs - stage2_num_epochs,
        switch_pipeline=train_pipeline_stage2)
]

# evaluators
val_evaluator = dict(
    type='CocoMetric',
    ann_file=data_root + 'annotations/annotations_coco_valid.json')
test_evaluator = dict(
    type='CocoMetric',
    ann_file=data_root + 'annotations/annotations_coco_valid.json')

The training went well, but I receive only preds for the first class using the inferencer:

image_list = get_list_of_images_array('/home/nico/PycharmProjects/opencv/yolov7_multiclass/DG_dataset_01/valid/')
img_path = '/home/nico/PycharmProjects/opencv/yolov7_multiclass/DG_dataset_01/valid/'   # replace this with your own image path

# create the inferencer using the model alias
inferencer = MMPoseInferencer(
    pose2d='/home/nico/PycharmProjects/mmpose/privat_configs/rtmpose-m_8xb64-210e-DG_dataset_01-256x256.py',
    pose2d_weights='/home/nico/PycharmProjects/mmpose/tools/work_dirs/rtmpose-m_8xb64-210e-DG_dataset_01-256x256/best_coco_AP_epoch_410.pth',
det_cat_ids = [1,2,3]
)

# The MMPoseInferencer API employs a lazy inference approach,
# creating a prediction generator when given input
result_generator = inferencer(image_list, show=False, out_dir = "/home/nico/PycharmProjects/mmpose/work_dirs/rtmpose-m_8xb64-210e-DG_dataset_01-256x256/inferencer")
result = next(result_generator)

I checked in the training pipeline what the dataloader delivers to the model and all annotations from all classes are in the dataloader. 

I also discovered with my debugger that the argument det_cat_ids = [1,2,3] that I pass to the inferencer is changed in the inference pipeline to (15, 16, 17, 18, 19, 20, 21, 22, 23) which is read from the file default_det_model.py. I have the suspicious that this could cause the problem, but not sure as the pipeline is quite complex. 

I would highly appreciate any suggestions or hints!

Best regards,
Nico

**Comment by Ben-Louis** [comment]
Thanks for using MMPose. If you use the inferencer without specifying the `det_model`, the default detection model and category IDs will be employed. You could try to specify `MMPoseInferencer(..., det_model='rtmdet-m', det_cat_ids = [1,2,3])` and check if the outputs match your expectation.

**Comment by NicolasPetermann134** [comment]
Many thx for the swift response! Unfortunately, it doesn't help. Do you have another idea, of what it might be?

**Comment by Ben-Louis** [comment]
Could you please specify the three classes that were used to train the RTMPose model?

**Comment by NicolasPetermann134** [comment]
That's the categories from my .json file:

[   {   'id': 1,
        'keypoints': [   'nose',
                         'left_eye',
                         'right_eye',
                         'left_ear',
                         'right_ear',
                         'left_shoulder',
                         'right_shoulder',
                         'left_elbow',
                         'right_elbow',
                         'left_wrist',
                         'right_wrist',
                         'left_hip',
                         'right_hip',
                         'left_knee',
                         'right_knee',
                         'left_ankle',
                         'right_ankle'],
        'name': 'person',
        'skeleton': [   [16, 14],
                        [14, 12],
                        [17, 15],
                        [15, 13],
                        [12, 13],
                        [6, 12],
                        [7, 13],
                        [6, 7],
                        [6, 8],
                        [7, 9],
                        [8, 10],
                        [9, 11],
                        [2, 3],
                        [1, 2],
                        [1, 3],
                        [2, 4],
                        [3, 5],
                        [4, 6],
                        [5, 7]],
        'supercategory': 'Defect'},
    {   'id': 2,
        'keypoints': [   'nose',
                         'left_eye',
                         'right_eye',
                         'left_ear',
                         'right_ear',
                         'left_shoulder',
                         'right_shoulder',
                         'left_elbow',
                         'right_elbow',
                         'left_wrist',
                         'right_wrist',
                         'left_hip',
                         'right_hip',
                         'left_knee',
                         'right_knee',
                         'left_ankle',
                         'right_ankle'],
        'name': 'barbell_front',
        'skeleton': [],
        'supercategory': 'Defect'},
    {   'id': 3,
        'keypoints': [   'nose',
                         'left_eye',
                         'right_eye',
                         'left_ear',
                         'right_ear',
                         'left_shoulder',
                         'right_shoulder',
                         'left_elbow',
                         'right_elbow',
                         'left_wrist',
                         'right_wrist',
                         'left_hip',
                         'right_hip',
                         'left_knee',
                         'right_knee',
                         'left_ankle',
                         'right_ankle'],
        'name': 'barbell_back',
        'skeleton': [],
        'supercategory': 'Defect'}]


An here an example of an image annot.:

[   {   'area': 24464,
        'bbox': [379.0, 668.0, 278.0, 88.0],
        'category_id': 3,
        'id': 1,
        'image_id': 0,
        'iscrowd': 0,
        'keypoints': [   523.2,
                         717.1,
                         2.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0],
        'num_keypoints': 17,
        'segmentation': []},
    {   'area': 86625,
        'bbox': [313.0, 721.0, 385.0, 225.0],
        'category_id': 2,
        'id': 2,
        'image_id': 0,
        'iscrowd': 0,
        'keypoints': [   508.62,
                         827.06,
                         2.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0,
                         0.0],
        'num_keypoints': 17,
        'segmentation': []},
    {   'area': 242913,
        'bbox': [273.0, 381.0, 561.0, 433.0],
        'category_id': 1,
        'id': 3,
        'image_id': 0,
        'iscrowd': 0,
        'keypoints': [   281.41,
                         462.93,
                         2.0,
                         295.92,
                         446.62,
                         2.0,
                         296.6,
                         460.78,
                         0.0,
                         358.67,
                         454.28,
                         2.0,
                         376.88,
                         447.13,
                         0.0,
                         433.75,
                         512.46,
                         2.0,
                         434.61,
                         512.61,
                         2.0,
                         493.55,
                         640.83,
                         2.0,
                         493.52,
                         642.42,
                         1.0,
                         515.65,
                         769.53,
                         2.0,
                         515.0,
                         745.48,
                         0.0,
                         800.53,
                         602.46,
                         2.0,
                         800.74,
                         603.36,
                         2.0,
                         475.35,
                         675.28,
                         2.0,
                         475.63,
                         677.96,
                         2.0,
                         603.73,
                         745.48,
                         0.0,
                         592.67,
                         738.33,
                         0.0],
        'num_keypoints': 17,
        'segmentation': []},
    {   'area': 34889,
        'bbox': [656.0, 297.0, 139.0, 251.0],
        'category_id': 1,
        'id': 4,
        'image_id': 0,
        'iscrowd': 0,
        'keypoints': [   683.46,
                         338.18,
                         2.0,
                         696.54,
                         327.76,
                         2.0,
                         694.2,
                         329.39,
                         0.0,
                         729.95,
                         332.31,
                         2.0,
                         713.7,
                         334.75,
                         0.0,
                         752.7,
                         377.65,
                         2.0,
                         720.85,
                         384.8,
                         2.0,
                         772.85,
                         435.18,
                         2.0,
                         708.5,
                         447.53,
                         2.0,
                         711.1,
                         475.15,
                         2.0,
                         665.6,
                         473.52,
                         2.0,
                         752.83,
                         521.0,
                         2.0,
                         727.45,
                         518.62,
                         2.0,
                         700.7,
                         581.75,
                         0.0,
                         661.7,
                         569.72,
                         0.0,
                         722.8,
                         620.42,
                         0.0,
                         683.15,
                         611.0,
                         0.0],
        'num_keypoints': 17,
        'segmentation': []}]


FYI: The categories 2 and 3 have actually only one kpt, why they have no skeleton and the remaining 16 kpts are always 0.


PS: sry for the bad formatting it don't apply my formatting for some reason.

**Comment by Ben-Louis** [comment, marked as answer]
The default object detector "RTMDet-m", which has been trained on COCO, unfortunately does not detect the 2nd and 3rd categories. To detect these objects, you may need to train a new object detector.

**Comment by NicolasPetermann134** [comment]
oh, I thought that within the training also the detector is trained:') Is there a guide to re-train such a detector "RTMDet-m" ? And where in the config file would I have to put the custom detector?

**Reply by Ben-Louis** [reply]
1. You can train an object detector with [MMDetection](https://github.com/open-mmlab/mmdetection)
2. Once the detector is trained, you can use it with MMPose inferencer following [this guide](https://mmpose.readthedocs.io/en/latest/user_guides/inference.html#custom-pose-estimation-models)

**Comment by NicolasPetermann134** [comment]
Many many thx for your help!
